### PR TITLE
Implement PSR-4-compliant namespacing

### PR DESCRIPTION
**Pull request by jugglinmike**
During my initial research into introducing an ORM for the "ae-api" project, I encountered a global function naming collision. [We previously opened an issue to use the PSR-4 convention in that project](https://github.com/wfmu/ae-api/issues/256), so I decided that it was appropriate to use it here, as well.

Unfortunately, most (if not all) of the code in this project may be removed soon. But since the "ae-auth" project still depends on it (and since it was helpful to trial this change with a smaller project), I went forward with this plan anyway.

This is a breaking change for both consumers of this framework. When/if this patch lands, I will follow up by publishing a new version and updating each project to consume it appropriately.

### CODE REVIEW

**Comment by jugglinmike** [comment]
I can't reproduce the errors reported by TravisCI. I'm going to try to debug, but if anyone else could try this out, I'd be curious to learn what happens on another not-CI machine.

**Comment by jugglinmike** [comment]
Nevermind. After provisioning a fresh VM, I can reproduce this locally. I got some splainin to do.

**Comment by jugglinmike** [comment]
Haha! Fixed it!

File: src/routes/BaseRoute.php [hunk 1]
```diff
@@ -1,4 +1,8 @@
 <?php
+namespace AEFramework\routes;
+
+use Exception;
+use UnexpectedValueException;
```
Comment (deathbearbrown) [inline src/routes/BaseRoute.php]: Do you need to add use for every type of exception in the file?

File: src/tests/BaseServiceTest.php [hunk 2]
```diff
@@ -1,6 +1,10 @@
 <?php
+namespace AEFramework\tests;
 
-abstract class BaseServiceTest extends PHPUnit_Extensions_Database_TestCase {
+use PDO;
+use Exception;
+
+abstract class BaseServiceTest \extends PHPUnit_Extensions_Database_TestCase {
```
Comment (deathbearbrown) [inline src/tests/BaseServiceTest.php]: is that \ supposed to be in front of extends or PHPUnit...

File: test/AEFrameworkTestCase.php [hunk 3]
```diff
@@ -1,10 +1,15 @@
 <?php
+namespace AEFramework\test;
+use PDO;
+use Exception;
+
 define('DB_HOST', '127.0.0.1');
 define('DB_USERNAME', 'api');
 define('DB_PASSWORD', 'bocoup');
 define('TEST_DB_DATABASE', 'api');
 define('DB_DATABASE', 'api');
```
Comment (deathbearbrown) [inline test/AEFrameworkTestCase.php]: shouldn't this use a testdb?

File: ansible/roles/composer/tasks/main.yml [hunk 4]
```diff
@@ -16,7 +16,7 @@
   tags: composer
 
 - name: install composer dependencies
-  composer: command=install working_dir={{root}}
+  composer: command=install working_dir={{root}} no_dev=no
```
Comment (bmac) [inline ansible/roles/composer/tasks/main.yml]: What does no_dev do?

File: src/routes/BaseRoute.php (hunk 1 above)
Comment (jugglinmike) [inline src/routes/BaseRoute.php]: Unfortunately, it looks like it's either that or prefix each reference with a `\`.

- https://secure.php.net/manual/en/language.namespaces.basics.php#116002
- https://secure.php.net/manual/en/language.exceptions.php#97963

File: src/tests/BaseServiceTest.php (hunk 2 above)
Comment (jugglinmike) [inline src/tests/BaseServiceTest.php]: Definitely `PHPUnit_Extensions_Database_TestCase`. I don't know how this is passing.

File: test/AEFrameworkTestCase.php (hunk 3 above)
Comment (jugglinmike) [inline test/AEFrameworkTestCase.php]: Maybe! But I'd like to avoid changing unrelated code for this patch.

**Comment by deathbearbrown** [comment]
Holy cannoli the tests run much faster.

File: ansible/roles/composer/tasks/main.yml (hunk 4 above)
Comment (jugglinmike) [inline ansible/roles/composer/tasks/main.yml]: `"Disables installation of require-dev packages ( see --no-dev )"](https://docs.ansible.com/ansible/composer_module.html). In this case, it's necessary to ensure that Composer initializes the autoloader correctly during the package installation process (see the "autoload-dev" field I'm adding to the `composer.json` file)

File: src/routes/BaseRoute.php (hunk 1 above)
Comment (deathbearbrown) [inline src/routes/BaseRoute.php]: are the regular SPL ones like InvalidArgumentException free?

Comment (jugglinmike) [inline src/routes/BaseRoute.php]: I doubt it, since it's required even for the basic `Exception` class.

**Comment by bmac** [comment]
Looks good to me. Tests pass locally. :+1:

**Comment by deathbearbrown** [comment]
+1 from me too. Tested locally

**Comment by jugglinmike** [comment]
Thanks, @bmac! I just caught a problem with some untested code--`Router.php` and `RouterInterface.php`. `ae-api` don't need these files, but `ae-auth` still requires them (for the time being, at least), so I've pushed up a fix.

**Comment by jugglinmike** [comment]
And thank you, also, @deathbearbrown!

**Comment by deathbearbrown** [comment]
Yeah but Auth will get refactored to use slim too. Possibly this week if I can get my VM to stop being dumb.

**Comment by jugglinmike** [comment]
Merged to `master` at commits:

- 70c4aae90cfb69e069bef9f6e9e4e4df0c107009
- 02c5b7a261e65f8402a1ae57b5ac6485ec6fc566

Thanks, you two!
